File: backends/ts-hono/src/test/auth.test.ts
import { describe, test, expect } from "vitest";
import { createApp } from "../app.js";
import { makeStaff, makeInvitation, signStaffCookie } from "./helpers.js";
import { signStaffId } from "../lib/staffSession.js";

const app = createApp();

describe("Auth", () => {
  describe("GET /api/auth/me", () => {
    test("認証済みでプロフィールが取得できる", async () => {
      const staff = await makeStaff();
      const res = await app.request("/api/auth/me", {
        headers: { Cookie: `staff_id=${signStaffCookie(staff.id)}` },
      });
      expect(res.status).toBe(200);
      const body = await res.json() as Record<string, unknown>;
      expect(body.staff_id).toBe(staff.id);
      expect(body.name).toBe(staff.name);
    });

    test("未認証で401が返る", async () => {
      const res = await app.request("/api/auth/me");
      expect(res.status).toBe(401);
    });

    test("署名の無い改ざんクッキーでは401が返る", async () => {
      const staff = await makeStaff();
      const res = await app.request("/api/auth/me", {
        headers: { Cookie: `staff_id=${staff.id}` },
      });
      expect(res.status).toBe(401);
    });

    test("署名部分が改ざんされたクッキーでは401が返る", async () => {
      const staff = await makeStaff();
      const signed = signStaffCookie(staff.id);
      const [idPart, expPart, sig] = signed.split(".");
      const tampered = `${idPart}.${expPart}.${sig.slice(0, -1)}${sig.at(-1) === "0" ? "1" : "0"}`;
      const res = await app.request("/api/auth/me", {
        headers: { Cookie: `staff_id=${tampered}` },
      });
      expect(res.status).toBe(401);
    });

    test("別のシークレットで署名されたクッキーでは401が返る", async () => {
      const staff = await makeStaff();
      const forged = signStaffId(staff.id, "wrong-secret", 3600);
      const res = await app.request("/api/auth/me", {
        headers: { Cookie: `staff_id=${forged}` },
      });
      expect(res.status).toBe(401);
    });

    test("有効期限切れのクッキーでは401が返る", async () => {
      const staff = await makeStaff();
      const secret = process.env.STAFF_COOKIE_SECRET ?? "test-staff-cookie-secret";
      const expired = signStaffId(staff.id, secret, -1);
      const res = await app.request("/api/auth/me", {
        headers: { Cookie: `staff_id=${expired}` },
      });
      expect(res.status).toBe(401);
    });
  });

  describe("GET /api/auth/logout", () => {
    test("ログアウトが成功する", async () => {
      const res = await app.request("/api/auth/logout");
      expect(res.status).toBe(200);
    });
  });

  describe("GET /api/auth/invitation/:token", () => {
    test("有効なトークンで招待情報が取得できる", async () => {
      const inv = await makeInvitation("valid-test-token");
      const res = await app.request(`/api/auth/invitation/${inv.token}`);
      expect(res.status).toBe(200);
      const body = await res.json() as Record<string, unknown>;
      expect(body.token).toBe(inv.token);
    });
  });

  describe("OAuth state nonce", () => {
    test("認可開始時にnonceクッキーを発行しstateに含める", async () => {
      const res = await app.request("/auth/google/redirect?token=inv-token");
      expect(res.status).toBe(302);
      const setCookie = res.headers.get("set-cookie") ?? "";
      const m = setCookie.match(/oauth_state=([^;]+)/);
      expect(m).not.toBeNull();
      expect(setCookie).toMatch(/HttpOnly/i);
      const nonce = m![1];
      const loc = new URL(res.headers.get("location") ?? "");
      expect(loc.searchParams.get("state")).toBe(`ts|${nonce}|inv-token`);
    });

    test("nonceクッキーが無い場合は400エラーページへリダイレクトする", async () => {
      const res = await app.request("/auth/google/callback?code=x&state=ts%7Cabc");
      expect(res.status).toBe(302);
      expect(res.headers.get("location")).toMatch(/\/error\?code=400$/);
    });

    test("nonceが一致しない場合は400エラーページへリダイレクトする", async () => {
      const res = await app.request("/auth/github/callback?code=x&state=ts%7Cabc", {
        headers: { Cookie: "oauth_state=xyz" },
      });
      expect(res.status).toBe(302);
      expect(res.headers.get("location")).toMatch(/\/error\?code=400$/);
    });

    test("stateにnonceセグメントが無い場合は400エラーページへリダイレクトする", async () => {
      const res = await app.request("/auth/github/callback?code=x&state=ts", {
        headers: { Cookie: "oauth_state=abc" },
      });
      expect(res.status).toBe(302);
      expect(res.headers.get("location")).toMatch(/\/error\?code=400$/);
    });
  });
});

File: backends/go-beego/tests/auth_test.go
package tests

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"testing"
)

func TestAuth_GetMyProfile(t *testing.T) {
	truncateTables(t)

	t.Run("認証済みでプロフィールが取得できる", func(t *testing.T) {
		staff := createStaff(t, nil)
		w := do(http.MethodGet, "/api/auth/me", nil,
			withCookie("staff_id", fmt.Sprintf("%d", staff.ID)))
		if w.Code != http.StatusOK {
			t.Errorf("want 200, got %d: %s", w.Code, w.Body.String())
		}
		body := parseBody(w)
		if body["staff_id"] == nil {
			t.Error("staff_id not found in response")
		}
		if body["name"] == nil {
			t.Error("name not found in response")
		}
	})

	t.Run("未認証で401が返る", func(t *testing.T) {
		w := do(http.MethodGet, "/api/auth/me", nil)
		if w.Code != http.StatusUnauthorized {
			t.Errorf("want 401, got %d", w.Code)
		}
	})
}

func TestAuth_Login(t *testing.T) {
	truncateTables(t)

	t.Run("認証済みでログイン情報が取得できる", func(t *testing.T) {
		staff := createStaff(t, nil)
		w := do(http.MethodGet, "/api/auth/login", nil,
			withCookie("staff_id", fmt.Sprintf("%d", staff.ID)))
		if w.Code != http.StatusOK {
			t.Errorf("want 200, got %d: %s", w.Code, w.Body.String())
		}
		body := parseBody(w)
		if body["staff_id"] == nil {
			t.Error("staff_id not found in response")
		}
	})

	t.Run("未認証で401が返る", func(t *testing.T) {
		w := do(http.MethodGet, "/api/auth/login", nil)
		if w.Code != http.StatusUnauthorized {
			t.Errorf("want 401, got %d", w.Code)
		}
	})
}

func TestAuth_Logout(t *testing.T) {
	t.Run("ログアウトが成功する", func(t *testing.T) {
		w := do(http.MethodGet, "/api/auth/logout", nil)
		if w.Code != http.StatusOK {
			t.Errorf("want 200, got %d: %s", w.Code, w.Body.String())
		}
	})
}

func TestAuth_Invitation(t *testing.T) {
	truncateTables(t)

	t.Run("有効なトークンで招待情報が取得できる", func(t *testing.T) {
		inv := createInvitation(t, "valid-test-token")
		w := do(http.MethodGet, "/api/auth/invitation/"+inv.Token, nil)
		if w.Code != http.StatusOK {
			t.Errorf("want 200, got %d: %s", w.Code, w.Body.String())
		}
		body := parseBody(w)
		if body["token"] == nil {
			t.Error("token not found in response")
		}
	})
}

func TestAuth_OAuthState(t *testing.T) {
	t.Run("認可開始時にnonceクッキーを発行しstateに含める", func(t *testing.T) {
		w := do(http.MethodGet, "/auth/google/redirect?token=inv-token", nil)
		if w.Code != http.StatusTemporaryRedirect {
			t.Fatalf("want 307, got %d", w.Code)
		}
		var nonce string
		for _, c := range w.Result().Cookies() {
			if c.Name == "oauth_state" {
				nonce = c.Value
				if !c.HttpOnly {
					t.Error("oauth_state cookie must be HttpOnly")
				}
			}
		}
		if nonce == "" {
			t.Fatal("oauth_state cookie not set")
		}
		loc, err := url.Parse(w.Header().Get("Location"))
		if err != nil {
			t.Fatal(err)
		}
		want := testCfg.OAuth.Runtime + "|" + nonce + "|inv-token"
		if got := loc.Query().Get("state"); got != want {
			t.Errorf("state: want %q, got %q", want, got)
		}
	})

	t.Run("nonceクッキーが無い場合は400エラーページへリダイレクトする", func(t *testing.T) {
		w := do(http.MethodGet, "/auth/google/callback?code=x&state=go-beego%7Cabc", nil)
		if w.Code != http.StatusTemporaryRedirect {
			t.Fatalf("want 307, got %d", w.Code)
		}
		if loc := w.Header().Get("Location"); !strings.HasSuffix(loc, "/error?code=400") {
			t.Errorf("want redirect to /error?code=400, got %s", loc)
		}
	})

	t.Run("nonceが一致しない場合は400エラーページへリダイレクトする", func(t *testing.T) {
		w := do(http.MethodGet, "/auth/google/callback?code=x&state=go-beego%7Cabc", nil,
			withCookie("oauth_state", "xyz"))
		if loc := w.Header().Get("Location"); !strings.HasSuffix(loc, "/error?code=400") {
			t.Errorf("want redirect to /error?code=400, got %s", loc)
		}
	})

	t.Run("stateにnonceセグメントが無い場合は400エラーページへリダイレクトする", func(t *testing.T) {
		w := do(http.MethodGet, "/auth/google/callback?code=x&state=go-beego", nil,
			withCookie("oauth_state", "abc"))
		if loc := w.Header().Get("Location"); !strings.HasSuffix(loc, "/error?code=400") {
			t.Errorf("want redirect to /error?code=400, got %s", loc)
		}
	})
}
